File: algorithms/2_medium/V/00794_valid_tic_tac_toe_state/solution.cpp
class Solution {
    inline bool check_win(vector<string>& b, size_t n, char player) {
        for (size_t i{0}; i < n; ++i) {
            bool is_winning{true};

            for (size_t j{0}; j < n; ++j) {
                if (b[i][j] != player) {
                    is_winning = false;
                    break;
                }
            }

            if (is_winning) {
                return true;
            }
        }

        for (size_t j{0}; j < n; ++j) {
            bool is_winning{true};

            for (size_t i{0}; i < n; ++i) {
                if (b[i][j] != player) {
                    is_winning = false;
                    break;
                }
            }

            if (is_winning) {
                return true;
            }
        }

        bool is_winning{true};;
        for (size_t i{0}; i < n; ++i) {
            if (b[i][i] != player) {
                is_winning = false;
                break;
            }
        }

        if (is_winning) {
            return true;
        }

        for (size_t i{0}; i < n; ++i) {
            if (b[i][n - 1 - i] != player) {
                return false;
            }
        }

        return true;
    }
public:
    bool validTicTacToe(vector<string>& board) {
        const size_t n{board.size()};
        int balance{0};

        for (size_t i{0}; i < n; ++i) {
            for (size_t j{0}; j < n; ++j) {
                if (board[i][j] == 'X') {
                    balance += 1;
                } else if (board[i][j] == 'O') {
                    balance -= 1;
                }
            }
        }

        if (balance > 1 || balance < 0) {
            return false;
        };

        bool is_player1_winner{check_win(board, n, 'X')};
        bool is_player2_winner{check_win(board, n,'O')};

        if (
            (is_player1_winner && is_player2_winner) ||
            (is_player1_winner && balance == 0) ||
            (is_player2_winner && balance == 1)
        ) {
            return false;
        }

        return true;
    }
};

File: algorithms/2_medium/V/00794_valid_tic_tac_toe_state/solution.rs
impl Solution {
    #[inline]
    fn check_win(b: &Vec<&[u8]>, n: usize, player: u8) -> bool {
        for i in 0..n {
            let mut is_winning = true;

            for j in 0..n {
                if b[i][j] != player {
                    is_winning = false;
                    break;
                }
            }

            if is_winning {
                return true;
            }
        }

        for j in 0..n {
            let mut is_winning = true;

            for i in 0..n {
                if b[i][j] != player {
                    is_winning = false;
                    break;
                }
            }

            if is_winning {
                return true;
            }
        }

        let mut is_winning = true;;
        for i in 0..n {
            if b[i][i] != player {
                is_winning = false;
                break;
            }
        }

        if is_winning {
            return true;
        }

        for i in 0..n {
            if (b[i][n - 1 - i] != player) {
                return false;
            }
        }

        return true;
    }
    pub fn valid_tic_tac_toe(board: Vec<String>) -> bool {
        let n = board.len();
        let board_bytes = board.iter().map(|x| x.as_bytes()).collect::<Vec<_>>();
        let mut balance = 0_i32;

        for i in 0..n {
            for j in 0..n {
                if board_bytes[i][j] == b'X' {
                    balance += 1;
                } else if board_bytes[i][j] == b'O' {
                    balance -= 1;
                }
            }
        }

        if balance > 1 || balance < 0 {
            return false;
        };

        let is_player1_winner = Self::check_win(&board_bytes, n, b'X');
        let is_player2_winner = Self::check_win(&board_bytes, n, b'O');

        if (
            (is_player1_winner && is_player2_winner) || 
            (is_player1_winner && balance == 0) ||
            (is_player2_winner && balance == 1)
        ) {
            return false;
        }

        true
    }
}
